coding/performance.md

```diff
@@ -0,0 +1,51 @@
+---
+title: Performance
+description: Ensuring what we build is highly performant
+---
+
+We have a responsibility to ensure that the products and services we provide on the web perform well.
+
+Many of our users consume our products in parts of the world where connectivity is poor, latency is high and data is expensive.
+
+We should treat performance as design - it's not an optional enhancement.
+
+## Performance considerations
+
+### Use images responsibly
+
+Images are one of the leading causes of bloated web pages. Techniques we should consider to mitigate their cost include;
+
+- Compress and optimise all images before reaching production. Optimising jpegs can remove up to 75% of their original file size.
+- Serve appropriately sized images proportionate to where the browser will display them.
+- Use appropriate file types. If possible, use WebP - a modern image format that provides superior lossless and lossy compression for images on the web.
+- Use SVGs where possible for vector based images.
+- Lazy load images that are not visible on initial page load.
+
+### Third party code
+
+Third party code should only be added after a careful cost benefit analysis. If you are including an externally linked library, you should be extra careful as the size of that library is outside our control.
+
+### Rendering
+
+How a web page renders has a huge effect on the perception of how responsive that page is - even small delays of milliseconds can degrade the user experience.
+
+- Wherever possible, defer the loading of third-party code until after `window.onload`
+- Where the above is not possible, use the `async` attribute on the `<script>` element to prevent blocking.
+- Animations and transforms should be done using CSS instead of Javascript.
+- Don't hijack native browser scroll
+- Limit DOM nodes to a minimum
+- Consider inlining critical CSS
+
+### Reduce HTTP requests
+
+Round trips to the server are expensive. We should be aware that our users may lose connection before they have downloaded all the assets required to complete page load. Reduce HTTP requests as much as possible by concatenating CSS & JS and embedding images and icons using SVG or Base64 formats.
```

Comment:
This isn't always the best case scenario. Sometimes splitting up the file to take advantage of multiple simultaneous requests is more efficient. Embedding images and icons also reduces the ability for the browser to cache the page making subsequent visits or other page visits slower.

Reply:
It was my understanding that this was only a consideration once we start utilising HTTP/2, as per:

https://developers.google.com/web/fundamentals/performance/get-started/httprequests-5

@hatched Are you saying we should remove this advice entirely or reword it?

Reply:
I think I'd propose removing it as it's a case-by-case decision.

Reply:
We are using HTTP/2 for most of our sites - everything except www.ubuntu.com, www.canonical.com, partners.ubuntu.com and assets.ubuntu.com. (Although these properties are very important). But we don't yet have HPush for any of them.

However, HTTP/2 only pushes the needle a bit in terms of the calculus. With HTTP/1 there's still a cutoff of size where assets above that size are better off being sent separately for their cacheabilty, and below, they're better being inlined, as the latency associated with the HTTP handshake is probably greater than the transfer time.

We could probably argue until the cows come home about where to put that threshold - and it very much depends on your connection speed / distance from the server. If we care a lot about those in the third world, we should actually be prioritising splitting & cacheability over inlining more.

But my instincts would be that the line falls somewhere between 100-200KB for HTTP/1, and maybe more like 50-100KB for HTTP/2.